README.md
<!-- SPDX-License-Identifier: GPL-2.0-only -->
<!--
/README.md

Copyright (C) 2022-2022  Akash Nag
-->

![ash logo](./assets/banner.png)

[![GPL-2.0 License](https://img.shields.io/badge/license-GPL_2.0-blue.svg)](https://github.com/akashnag/ash/blob/master/LICENSE.txt) ![Downloads](https://img.shields.io/github/downloads/akashnag/ash/total) ![Size](https://img.shields.io/github/size/akashnag/ash/dist/ash-editor-0.1.0.dev14.tar.gz) ![SLOC](https://sloc.xyz/github/akashnag/ash)

**ash** is a simple and clean terminal-based text editor, that aims to be easy to use with modern key-bindings. It is capable of handling multiple files simultaneously and has a wide array of modern features. Here is a picture of **ash** editing this README file:

![Screenshot](./assets/ash-default.png)

**ash** is written in Python 3.8 using the curses library.

## Table of Contents

- [Features](#features)
- [Release Notes](RELEASE_NOTES.md)
- [Requirements](#requirements)
- [Installation](#installation)
  - [Prerequisites](#prerequisites)
  - [Installing **ash**](#installing-ash)
  - [Uninstalling **ash**](#uninstalling-ash)
  - [Setting paths](#setting-paths)
- [Prebuilt binaries](#prebuilt-binaries)
- [Customization](CUSTOMIZATION.md)
- [Usage](#usage)
- [Contributing](#contributing)
- [Screenshots](#screenshots)
- [License](#license)

- - -

## Features

The following is a list of features available in **ash**:

- Easy to use, clean and intuitive interface
- Common key bindings (Help on F1, Arrow keys for movement, Shift+Arrow/Home/End/PgUp/PgDown for selecting text, cut/copy/paste using Ctrl+X/C/V, undo/redo using Ctrl+Z/Y, find/replace/goto using Ctrl+F/H/G, etc.) and support for remapping key bindings to your taste
- Common editor features such as undo/redo, line numbers, find-replace, cut-copy-paste, etc.
- True support for wrapping (both hard & soft) with intuitive cursor movements along wrapped text
- Multi-file editing with support for windows and tabs
- Auto-backup
- Support for Unicode
- Project mode (opening a directory instead of individual files)
- Complete session (for projects) and undo persistence (turned on for projects opened directly from command-line)
- Live search
- Support for search/replace in **all** open files
- Support for searching using regular expressions
- Auto-indentation, Select+Tab/Shift-Tab to increase/decrease indent
- Auto insertion of matching braces/quotes and auto-enclosure when text is selected and braces/quotes are typed
- Support for unlimited splits per tab (subject to screen size) and support for unlimited tabs
- Support for various text-encodings
- Checks (live) and reloads (if user permits) files which have been modified externally
- Selection highlighting (highlights text under selection wherever they occur in the document)
- Color scheme customization
- View list of recent files, view project explorer (in project mode)
- Syntax highlighting (limited)
- Git integration (shows untracked, modified files, etc.)
- Multiple Cursors
- Command palette
- Basic mouse support (click to select, right-click, drag to select text, double-click to select word, triple-click to select line, drag to move dialog-boxes, etc.)
- Open terminal / execute shell commands
- Limited IDE features like compile, build, execute...
- Supports project/folder specific settings along with global settings
- Multi-language support
- Support for snippets

## Requirements

1. You need a resolution of at least 102 x 22 in your terminal emulator
1. Your terminal must support Unicode and be able to display at least 256 colors with the capability of remapping color palettes (Works best on: `xterm-256`). Monochrome terminals are also partially supported.
1. If you are running the source distribution, you need the GNU C compiler collection besides Python 3.8, as some parts of the application are written in Cython. The first time you run the application, Cython will compile and build the `*.pyx` files, which may increase load time (Ignore the warnings during compilation)

**ash is primarily targetted to Linux. This version of ash has only been tested on Ubuntu 20.04 with Python 3.8.2 (64-bit)**

## Installation

*Note: Instructions provided here are for Debian/Ubuntu based systems. For your distribution, you may need to issue different instructions to install the necessary packages.*

### Prerequisites

You need certain packages and Python 3 itself to run **ash** on your system, as there are no prebuilt binaries available yet.

```bash
$ sudo apt install git
$ sudo apt install xclip
$ sudo apt install python3
$ sudo apt install python3-pip
```

Depending on your system, you may also need to install the following packages:

```bash
$ sudo apt install python3-devel
```

### Installing ash

You have multiple options here:

- Install from PyPi **(recommended)**
- Install the latest stable release from GitHub
- Install the latest nightly release from GitHub
- Build and install from source

#### Install from PyPi

Execute the following instructions to install from PyPi:

```bash
$ sudo pip3 install ash-editor
```

The above command downloads **ash** and installs it locally on your system. To be able to invoke **ash** from anywhere, see the [Setting Paths](#setting-paths) section.

#### Install the latest stable release

**Note: the stable release may not contain all features listed above**

Go to the **ash** [Website](https://akashnag.github.io/ash) and download the latest stable release. Then extract the downloaded tar-ball into a folder and once inside that folder, open up your terminal from there and execute:

```bash
$ sudo pip3 install .
```

#### Install the latest nightly release

The nightly builds contains all the latest features of ash. Go to the **ash** [Website](https://akashnag.github.io/ash) and download the latest nightly release. Then extract the downloaded tar-ball into a folder and once inside that folder, open up your terminal from there and execute:

```bash
$ sudo pip3 install .
```

#### Build from source and install

To build from source, issue the following instructions:

```bash
$ sudo apt install git
$ cd ~
$ git clone --depth 1 -b master https://github.com/akashnag/ash.git
$ cd ash
$ python3 setup.py sdist
```

(Though you can install it directly from here, it is not recommended) You will find the new tar-ball created under `dist` directory, from where you can proceed as before by first copying the tarball into a new directory, extracting it, and then installing it.


### Setting paths

To run **ash** make sure you have `:$HOME/.local/bin` appended to your $PATH variable in the file `~/.bashrc` (If you are using `bash` as your shell. If not open your corresponding .rc file). To execute **ash**, see the [Usage](#usage) section.

### Uninstalling ash

To uninstall **ash** you can use:

```bash
$ sudo pip3 uninstall ash-editor
```

## Prebuilt binaries

Since **ash** is still under development, prebuilt binaries are not yet available. You can use PyInstaller or similar tools to build one for your system.


## Usage

Once you have downloaded the **ash** source code, and set it up as detailed above, you are ready to use it.

**NOTES:**

1. If you have not updated your path variable, you must specify the full path to the ash binary.
1. Your terminal resolution should be at least 102 (width) x 22 (height). Opening the editor in a lower resolution may unexpectedly crash the application. This requirement is necessary to properly display the dialog-boxes.
1. It will take a while to start up **ash** for the first time, this is normal. From the second time onwards, it should start up faster.

To run **ash**:

```bash
$ ash path/to/file.txt
```

or, to open an empty buffer:

```bash
$ ash
```

or, to open a project (directory):

```bash
$ ash path/to/directory
```

See the [Key Bindings](KEYBINDINGS.md) for help on how to navigate in ash.

## Contributing

### Contributing to the source code

If you find any bugs, please report them. I am also happy to accept pull requests from anyone for either bug-fixes, performance improvements, or for implementing the not-yet-implemented features listed above. Please consider contributing towards new features **only when** the features listed above have been fully implemented. For more information on what features to implement and the project structure, see the [Contribution Guidelines](CONTRIBUTING.md)

### Bug-fixes and feature enhancements

You can use the [GitHub issue tracker](https://github.com/akashnag/ash/issues) to report bugs, ask questions, or suggest new features. You can include the log if that is relevant, which can be found in `~/.config/ash-editor/log.txt`.

### Discussions

For discussions related to the development roadmap and the **ash** editor in general, you can join the [ash Discord Server](https://discord.gg/KErp57T6en).

### Say Thanks\!

If you are not into software development, you can still contribute financially. Financial contributions help me prioritize working on this project over others and helps me know that there is a real demand for this software.

<a href="https://www.buymeacoffee.com/nagakash"><p align="center"><img src="./assets/bmc-button.png" height="70"/></p></a>


## Screenshots

![Screenshot](./assets/ss1.png)
![Screenshot](./assets/ss2.png)
![Screenshot](./assets/ss3.png)
![Screenshot](./assets/ss4.png)
![Screenshot](./assets/ss5.png)
![Screenshot](./assets/ss6.png)
![Screenshot](./assets/ss7.png)
![Screenshot](./assets/ss8.png)
![Screenshot](./assets/ss9.png)
![Screenshot](./assets/ss10.png)
![Screenshot](./assets/ss11.png)
![Screenshot](./assets/ss12.png)
![Screenshot](./assets/ss13.png)
![Screenshot](./assets/ss14.png)
![Screenshot](./assets/ss15.png)
![Screenshot](./assets/ss16.png)
![Screenshot](./assets/ss17.png)
![Screenshot](./assets/ss18.png)
![Screenshot](./assets/ss19.png)
![Screenshot](./assets/ss20.png)

## License

Copyright &copy; Akash Nag. All rights reserved.

Licensed under the [GPL-2.0](LICENSE.txt) license.


RELEASE_NOTES.md
<!-- SPDX-License-Identifier: GPL-2.0-only -->
<!--
/RELEASE_NOTES.md

Copyright (C) 2022-2022  Akash Nag
-->

# Release Notes

| Version | Release Date |
|---------|--------------|
| [0.1.0-dev14](#10th-august-2022-release) | 2022-08-10 |
| [0.1.0-dev13](#6th-august-2022-release) | 2022-08-06 |
| 0.1.0-dev12 | 2022-02-20 |
| 0.1.0-dev11 | 2022-02-06 |
| 0.1.0-dev10 | 2022-01-25 |


## 10th August 2022 Release

### New features

#### Multi-language support

From this release, **ash** now supports multiple languages. You can add language files (in JSON format) and extend the language support. By default, **ash** only supports `en-US`. To know more about how you can create and add language files, see [Adding language support](CUSTOMIZATION.md#adding-languages).

#### Support for Snippets

From this release, **ash** now supports snippets. You can add your own snippets and then insert them any where in your text. To add a defined snippet, select `Edit` > `Insert snippet`. In the dialog-box which opens you can type to search a snippet name. The list of snippets matching your search will appear in the list below. You can type the full snippet name and press Enter, or move to the list and select a snippet and press Enter to insert the snippet. Besides the snippet name, you can also type its numeric ID for a faster snippet selection. To know more about how you can create and add snippets to **ash**, see [Adding snippets](CUSTOMIZATION.md#adding-snippets).

#### Improved mouse support

Mouse support has been further improved. Now you can:

- Double-click to select a word, triple-click to select a line in the editor
- Move cursor up/down in the editor using the mouse-wheel
- Scroll up/down in lists in all dialog-boxes
- Drag the mouse to select text in the editor

#### Supported file types are now customizable

Before opening a file, **ash** checks its MIME-type to determine whether or not it is a binary or a text file. Binary files can not be opened in **ash**. The list of supported MIME-types were pre-defined. Also, **ash** did not take into consideration the file extension to determine whether or not it is a text file. This now changes. From this release, **ash** both MIME-types as well as file-extensions will be checked. And both of these lists (i.e. list of supported MIME types and list of supported file-extensions) are now customizable in settings.

### Bug fixes

#### Menus and close-buttons are now clickable

Due to a bug introducted in the previous version, menu-bar, popup-menus and close-buttons were rendered unclickable. Now the bug has been fixed.

#### Settings, theme files, etc. were missing the last line

JSON files created by **ash** were missing a line-feed character at the end, and the last line was not being displayed in the editor when those files were loaded. Now, **ash** will add a new line to the end of every file it creates. (Note: files created using other editors which do not add the newline at the end, will still be displayed incorrectly.)

#### Project Explorer search functionality fixed

The search functionality in Project Explorer was not able to perform correctly. This has now been fixed.

#### Invisible text in lists fixed

Due to a bug in the color module, some texts in list-boxes were rendered invisible (e.g. in Project Explorer and Find/Replace in all files). This has now been fixed.

#### Other bug fixes

Sometimes the editor crashed when the menu bar changed state from visible to hidden after selecting an item. This has now been fixed.

### Other changes

#### Adding themes and keymaps are now easier

Creating new themes and keymaps, and applying them are now straight-forward. To know how you can create and apply themes and keymaps, see [Customization](CUSTOMIZATION.md).

#### Menu revamped

The **ash** menu has been redesigned with some menu items shifted, some menus renamed, and others like `Close` and `Reload from disk` options added.

<!------------------------------------------------------------------------------------>

## 6th August 2022 Release

### New features

#### Improved mouse support

Mouse support has been improved. Clicking on the title-bar will now show the menu-bar. Right-clicking on a line of text in the editor will now show the context-menu. Clicking on an inactive editor will now move the focus to that editor. Clicking on any text will move the cursor to that position. Dialog-boxes can now be dragged using their title-bars.

### Bug fixes

#### Error on writing to write-protected file
	
The application would crash if a write-protected file were to be opened and then attempts are made to save to it. This is now resolved. An error message is now displated to the user if they try to save any changes to the file. Users can then try to save the file with a different name if they so choose.

#### File deleted externally

If a currently opened file is deleted externally, <b>ash</b> would now inform the user and let them choose what to do with the file. If they choose to recreate the file, it will be recreated. Otherwise, the contents will be transferred to a new unfiled buffer.

### Other changes

#### Settings moved to `~/.config` on Linux
	
<a href="https://github.com/domsch1988">@domsch1988</a><br/>

All settings, keymaps, theme files, etc. have now been moved from `~/.ash-editor` to `~/.config/ash-editor`. <a href="https://github.com/akashnag/ash/pull/17">PR #17</a>


src/ash/__init__.py
# SPDX-License-Identifier: GPL-2.0-only
#
# /src/ash/__init__.py
#
# Copyright (C) 2022-2022  Akash Nag

# This is imported by all modules, sets global parameters

import curses
import os
import sys
import copy
import glob
import subprocess
import platform

__version__			= "0.1.0-dev"
__build__			= "14"
__release_date__	= "Aug 10, 2022"

APP_COPYRIGHT_TEXT	= "© Copyright 2020-2022, Akash Nag. All rights reserved."
APP_LICENSE_TEXT	= "Licensed under the GPL-2.0 License."
APP_WEBSITE_URL		= "https://akashnag.github.io/ash"
APP_GITHUB_URL		= "https://github.com/akashnag/ash"

# list of supported encodings
SUPPORTED_ENCODINGS = sorted([ 
                            "utf-8", "utf-7", "utf-16", "utf-32", 
                            "latin-1", "tis-620"
                            "big5", "gb2312", "gb18030", "euc-tw", "hz-gb-2312", 
                            "iso-2022-cn", "euc-jp", "shift_jis", "iso-2022-jp",
                            "euc-kr", "iso-2022-kr", "koi8-r", "maccyrillic", "ibm855",
                            "ibm866", "iso-8859-5", "windows-1251", "iso-8859-2",
                            "windows-1250", "iso-8859-5", "iso-8859-1", "windows-1252",
                            "iso-8859-7", "iso-8859-8", "windows-1253", "windows-1255"                            
                        ])

APP_MODE_FILE		= 1		# if ash is invoked with zero or more file names
APP_MODE_PROJECT	= 2		# if ash is invoked with a single directory name

# dictionary of settings for global access
SETTINGS			= None

# minimum resolution required for ash
MIN_WIDTH			= 102
MIN_HEIGHT			= 22

# no. of milliseconds to sleep before checking for input (decreasing it will increase CPU usage; increasing it will reduce responsiveness)
SLEEP_MS            = 10

# minimum dimensions for any particular editor
MIN_EDITOR_WIDTH	= 15
MIN_EDITOR_HEIGHT	= 5

# path to the application data directory and configuration files
if platform.system() == "Linux":
    APP_DATA_DIR			= os.path.join(os.path.expanduser("~"), ".config", "ash-editor")
else:
    APP_DATA_DIR			= os.path.join(os.path.expanduser("~"), ".ash-editor")

APP_PLUGINS_DIR			= os.path.join(APP_DATA_DIR, "plugins")
APP_KEYMAPS_DIR			= os.path.join(APP_DATA_DIR, "keymaps")
APP_THEMES_DIR			= os.path.join(APP_DATA_DIR, "themes")
APP_LOCALES_DIR			= os.path.join(APP_DATA_DIR, "locales")
APP_SNIPPETS_DIR		= os.path.join(APP_DATA_DIR, "snippets")

PROJECT_SETTINGS_DIR_NAME  = ".ash-editor"
PROJECT_SETTINGS_FILE_NAME = "settings.json"

TEMP_OUTPUT_FILE		= os.path.join(APP_DATA_DIR, "temp.output")
LOG_FILE 				= os.path.join(APP_DATA_DIR, "log.txt")
SESSION_FILE			= os.path.join(APP_DATA_DIR, "session.dat")
SETTINGS_FILE			= os.path.join(APP_DATA_DIR, "settings.json")

DEFAULT_SHELL			= "bash"
DEFAULT_TERMINAL		= "xterm"

THEME_FILE_EXTENSION    = ".json"
KEYMAP_FILE_EXTENSION   = ".json"
LOCALE_FILE_EXTENSION   = ".json"
SNIPPET_FILE_EXTENSION  = ".snippets"